Persist trip cost via Mongo

marvindore opened the pull request:
There is at most one cost per user stored in DB. Existing trip cost are updated as the user creates more trips.

Commit:
Initial work on adding rabbitmq to calculation service

Commit:
A bunch of work on rabbitmq messaging

Commit:
Fixed an error in main.go

Commit:
Full chain of communication working

Commit:
refactor project structure and cleaned up functions that are no longer in use

Commit:
Fixed example post request in readme

Commit:
Generalized rabbitmq credentials

Commit:
Merged Master

Commit:
Resolve Merge conflicts

Commit:
Resolve merge conflicts

Commit:
Add mongodb connection and GET/POST endpoints

Integrate MongoDB to store all costs with
an endpoint which will return only the most
trip calculation.

Commit:
Each user should have only one cost/trip in DB

When user accepts a trip the cost is created
in the database if it doesn't exist. If there
already exists a trip for the user then the
existing trip is updated.

jschwait reviewed (changes requested):
Add all the .idea files to your .gitignore file so they're not added to your pushes

jschwait reviewed (changes requested):
Your files are failing the formatting checks.  Check out the "./.travis.gofmt.sh" file for what to fix

jschwait commented on db/db.go:
```diff
@@ -0,0 +1,71 @@
+//	var cost Cost
+//	err := MgoSession.C(COLLECTION).FindId(bson.ObjectIdHex(id)).One(&cost)
+//	return trip, err
+//}
```
If this code is unnecessary please remove the comments

jusoto replied:
code was removed

jschwait commented on handlers/rabbitmq_handlers.go:
```diff
@@ -59,6 +65,25 @@ func calculateCost(gmapsEstimation models.Estimation) models.Estimation {
+	cost.Destination = gmapsEstimation.Destination
+	cost.Cost = finalCost
+	//Write cost to db
+	//upsetdata := bson.M{"$set": cost}
```
is this comment necessary?

jusoto replied:
comment removed

Commit:
Include GET endpoint in README file

Commit:
Remove unintentional comments

Commit:
Ignore .idea files in git

jschwait commented on .gitignore:
```diff
@@ -1,2 +1,3 @@
+.idea
```
I think it has to be .idea/ so that all the files are ignored

jusoto replied:
Change made

jschwait commented on main.go:
```diff
@@ -2,16 +2,22 @@ package main
 
 import (
 	"github.com/AITestingOrg/calculation-service/handlers"
+	//"github.com/AITestingOrg/calculation-service/db"
```
unnecessary comment

jusoto replied:
change made

jschwait reviewed (dismissed):
Still failing the formatting check

Commit:
Refactor code and fixed formatting

jacobkissel commented on handlers/api_handlers.go:
```diff
@@ -3,27 +3,58 @@ package handlers
 	Publisher interfaces.PublisherInterface
 }
+type CostStorage struct {
+	Publisher interfaces.PublisherInterface
```
It looks like this CostStorage handler doesn't need to have a publisher because it doesn't need to publish anything as far as I can see.

jusoto replied:
Changed to ControllerInterface

jacobkissel commented on handlers/rabbitmq_handlers.go:
```diff
@@ -59,6 +64,24 @@ func calculateCost(gmapsEstimation models.Estimation) models.Estimation {
+	log.Println("Writing cost to database")
+	c := session.DB("TRIPCOST").C("costs")
+
+	_, err := c.Upsert(bson.M{"userId": cost.UserId}, cost)
```
Would it be possible to use the interface constructor for this, like cost = models.Cost{ Origin: gmapsEstimation.Origin, UserId: gmapsEstimation.UserId .... etc }

jusoto replied:
Created a file to manage Mongo sessions.

jacobkissel commented on handlers/api_handlers.go:
```diff
@@ -3,27 +3,58 @@ package handlers
 	body, _ := ioutil.ReadAll(r.Body)
-
-	//Converts objects into json
+	//	//Converts objects into json
```
Weird commenting

jusoto replied:
fixed

jacobkissel commented on main.go:
```diff
@@ -10,8 +11,12 @@ func main() {
 	amqpPublisher := new(utils.AmqpPublisher)
 
+	//close mongo session
+	defer db.MgoSession.Close()
```
Where does this db.MgoSession get opened? Its init function shows no usages. Is it implicit?

marvindore replied:
This closes the main mongo session in the db/db.go file.

Castyr replied:
dangling closes like this are dangerous, you should close in the same file the connection was opened in.

jusoto replied:
connection is opened and closed in the same file

jacobkissel commented on docker-compose-local.yml:
```diff
@@ -26,7 +26,7 @@ services:
       - discoveryservice
       - rabbitmq
-    image: aista/gmaps-adapter
+    build: ../gmaps-adapter
```
This convention doesn't really work for most people's file structure, because we have to have this project in the gopath, and other projects are normally somewhere else.

jusoto replied:
changed back to aista/gmaps-adapter

Commit:
Modify endpoint and Use controller interface instead of Publisher

Commit:
Refactor code to use constructor

Commit:
Remove time variable from struct and Publisher args in costStorage interface

Commit:
Use dockerhub images

Castyr commented on .idea/calculation-service.iml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Don't check .idea stuff in.

jusoto replied:
.idea added to .gitignore

Castyr commented on db/db.go:
```diff
@@ -0,0 +1,23 @@
+var MgoSession *mgo.Session
+
+func init() {
+	session, err := mgo.Dial("mongo:27017")
```
This should be configurable.

jusoto replied:
added environment variables MONGO_HOST for IP address, MONGO_PORT for port communication

Castyr commented on handlers/api_handlers.go:
```diff
@@ -16,13 +18,44 @@ type CostEstimateHandler struct {
+	log.Println("GET request UUID retreived: ", cost.UserId)
+	err := c.Find(bson.M{"userId": cost.UserId}).Limit(1).One(&cost)
+	if err != nil {
+		log.Printf("Failed to find cost")
```
This should respond with a 404.

jusoto replied:
added http 404 error

Commit:
Use environment variables for DB connection

jusoto committed:
Merge branch 'master' into mongodb-persistence

Commit:
Merge remote-tracking branch 'origin/master' into mongodb-persistence

Commit:
Update travis for RabbitMQ

Commit:
Merge branch 'mongodb-persistence' of github.com:AITestingOrg/calculation-service into mongodb-persistence

Commit:
Remove RabbitMQ from travis

jusoto committed:
Updated Trip Object to match the json objects being shared with other microservices "originAddress" "destinationAddress"

jusoto committed:
Update .travis.yml with environment variables and notify only on success to slack channel

jusoto committed:
Update travis environment variables with localhost

jusoto committed:
Removed log heartbeat from Eureka Client to make more readable the logs

jusoto committed:
Corrected originAddress to origin, destinationAddress to destination

jusoto committed:
Updates:
- Updated test cases to match Mongodb implementation
- Created a Data Access Interface to mock connection
- Updated fields to match object models with external services

jusoto committed:
Added auto format command for travis

jusoto committed:
Fix formatting
